### PR TITLE
Implement type hinting predicates for queries

### PR DESCRIPTION
**Pull request by thomcc**
I think the name "hint" is a misnomer (it's more of a type check for an existing item).

If it needs more tests let me know.

### CODE REVIEW

File: query-algebrizer/src/clauses/hints.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    ErrorKind,
+    Result,
+};
+
+use types::{
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+
```
Comment (rnewman) [inline query-algebrizer/src/clauses/hints.rs]: Nit: kill a few of these surplus newlines, here and elsewhere.

File: query-algebrizer/src/clauses/hints.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    pub fn apply_type_hint(&mut self, pred: Predicate, ty: ValueType) -> Result<()> {
+        if pred.args.len() != 1 {
+            bail!(ErrorKind::InvalidNumberOfArguments(pred.operator.clone(), pred.args.len(), 1));
+        }
+
+        if let FnArg::Variable(ref v) = pred.args[0] {
+            if let Some(existing) = self.type_hints.insert(v.clone(), ty) {
```
Comment (rnewman) [inline query-algebrizer/src/clauses/hints.rs]: I agree with your naming concern, and I think I would call this `required_types` instead of `type_hints`.

Comment (rnewman) [reply query-algebrizer/src/clauses/hints.rs]: Refactoring request: make this a separate method that takes a type and a variable. Leave the arg parsing to this function. We will use this functionality elsewhere (e.g., to fix #520).

Comment (rnewman) [reply query-algebrizer/src/clauses/hints.rs]: Change request: make `type_hints` (aka `required_types`) a `ValueTypeSet`, not a `ValueType`. That way we can impose both numeric types.

File: query-algebrizer/src/clauses/not.rs [hunk 3]
```diff
@@ -53,13 +53,14 @@ impl ConjoiningClauses {
             template.apply_clause(&schema, clause)?;
         }
 
+        template.expand_column_bindings();
+        template.prune_extracted_types();
+        template.process_required_types()?;
+
         if template.is_known_empty() {
```
Comment (rnewman) [inline query-algebrizer/src/clauses/not.rs]: Do this check before and after each step, no?

File: query-algebrizer/src/clauses/predicate.rs [hunk 4]
```diff
@@ -34,11 +34,26 @@ use types::{
...
+        _ => None
+    }
+}
+
 /// Application of predicates.
 impl ConjoiningClauses {
     /// There are several kinds of predicates in our Datalog:
     /// - A limited set of binary comparison operators: < > <= >= !=.
     ///   These are converted into SQLite binary comparisons and some type constraints.
+    /// - A set of type requirements constraining their argument to be a specific ValueType
```
Comment (rnewman) [inline query-algebrizer/src/clauses/predicate.rs]: Closing period.

File: query-algebrizer/src/types.rs [hunk 5]
```diff
@@ -334,11 +334,25 @@ pub enum ColumnConstraint {
...
+        value: TableAlias,
+        value_types: ValueTypeSet,
+        strict: bool,
+    },
     NotExists(ComputedTable),
     Matches(QualifiedAlias, QueryValue),
 }
 
+impl ColumnConstraint {
+    pub fn has_type(value: TableAlias, value_type: ValueType) -> ColumnConstraint {
```
Comment (rnewman) [inline query-algebrizer/src/types.rs]: TODO: expand this to a `ValueTypeSet`.

File: query-algebrizer/tests/utils/mod.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    algebrize_with_inputs,
+    ConjoiningClauses,
+    Error,
+    QueryInputs,
+};
+
+// Common utility functions used in multiple test files. Note: Import this with
+// `pub mod utils` (not `mod utils`), or you'll get spurious unused function
+// warnings when functions exist in this file but are only used by modules that
+// don't import with `pub` (yes, this is annoying).
```
Comment (rnewman) [inline query-algebrizer/tests/utils/mod.rs]: Better to add `#![allow(unused_code)]` underneath the block comment in this file, and not use `pub mod`.

File: query-sql/src/lib.rs [hunk 7]
```diff
@@ -345,11 +371,16 @@ impl QueryFragment for Constraint {
                     out.push_sql("0");
                     return Ok(())
                 }
-                out.push_sql("(");
+                let wrap_parens = constraints.len() > 1;
```
Comment (rnewman) [inline query-sql/src/lib.rs]: Skip the conditional; always output parens. It's safe and it avoids two branches.

File: query-translator/tests/translate.rs [hunk 8]
```diff
@@ -286,6 +286,50 @@ fn test_unknown_ident() {
...
+                            typeof(`all_datoms00`.v) = 'integer')");
+
+    assert_eq!(args, vec![]);
+}
+
+#[test]
+fn test_type_required_double() {
+    let schema = Schema::default();
+
+    let query = r#"[:find ?x :where [?x _ ?e] [(double ?e)]]"#;
```
Comment (rnewman) [inline query-translator/tests/translate.rs]: Here's a question for you. If we rephrase this as

`[:find ?x :where [(double ?e)] [?x _ ?e]]`

do we still use `all_datoms` rather than `datoms`? If so, please file a bug for that.

File: query-algebrizer/src/types.rs (hunk 5 above)
Comment (thomcc) [reply query-algebrizer/src/types.rs]: This is intended to be a helper for the most common case, e.g. the single type case where loose checks are desired.

File: query-algebrizer/src/clauses/mod.rs [hunk 9]
```diff
@@ -844,7 +876,43 @@ impl ConjoiningClauses {
...
+                }
+                if already_known.is_subset(types) {
+                    // We're certain the value is already constrained enough.
+                    continue;
+                }
+            }
 
+            let qa = self.extracted_types
+                .get(&var)
+                .ok_or_else(|| Error::from_kind(ErrorKind::UnboundVariable(var.name())))?;
```
[the remark this answers is not in the document]
Comment (thomcc) [reply query-algebrizer/src/clauses/mod.rs]: This doesn't work without additional type hinting, which seems to require an intermediate variable.

File: query-algebrizer/src/clauses/pattern.rs [hunk 10]
```diff
@@ -444,7 +444,7 @@ mod testing {
         // TODO: implement expand_type_tags.
         assert_eq!(cc.wheres, vec![
                    ColumnConstraint::Equals(d0_v, QueryValue::TypedValue(TypedValue::Boolean(true))),
-                   ColumnConstraint::HasType("datoms00".to_string(), ValueType::Boolean),
+                   ColumnConstraint::has_type("datoms00".to_string(), ValueType::Boolean),
```
Comment (rnewman) [inline query-algebrizer/src/clauses/pattern.rs]: Maybe call this `has_unit_type`?

File: query-translator/tests/translate.rs (hunk 8 above)
Comment (thomcc) [reply query-translator/tests/translate.rs]: Just so that it's recorded, this is no longer an issue as of b0d96e598a6f29e6dd8701279860e423d895ae72 (commit 4), which ensures that we use top level type hints when determining which table to use, even if they appear late.

File: query-algebrizer/src/clauses/mod.rs [hunk 11]
```diff
@@ -873,6 +968,26 @@ impl ConjoiningClauses {
 }
 
 impl ConjoiningClauses {
+    pub fn apply_clauses(&mut self, schema: &Schema, where_clauses: Vec<WhereClause>) -> Result<()> {
+        let mut deferred = Vec::with_capacity(where_clauses.len());
+        // We apply (top level) type predicates first as an optimization.
+        for c in where_clauses {
```
Comment (rnewman) [inline query-algebrizer/src/clauses/mod.rs]: If you're going to collect into `deferred`, then use `into_iter` here to consume the input `Vec`… and then you can even use a filter and `collect` to do it.

But I have a better suggestion. Can we just add a `WhereClause::TypeAnnotation` case to make your check efficient, and go back to walking once by reference and once by value? Maybe a `good-first-bug` follow-up, 'cos it adds a bit more work to something that already works.

File: query-algebrizer/src/clauses/predicate.rs [hunk 12]
```diff
@@ -34,11 +34,26 @@ use types::{
     Inequality,
 };
 
+pub fn parse_type_predicate(s: &str) -> Option<ValueType> {
```
Comment (rnewman) [inline query-algebrizer/src/clauses/predicate.rs]: In my suggestion, this would go into the query parser, I think.

File: query-algebrizer/src/types.rs [hunk 13]
```diff
@@ -334,11 +334,25 @@ pub enum ColumnConstraint {
         left: QueryValue,
         right: QueryValue,
     },
-    HasType(TableAlias, ValueType),
+    HasTypes {
+        value: TableAlias,
+        value_types: ValueTypeSet,
+        strict: bool,
```
Comment (rnewman) [inline query-algebrizer/src/types.rs]: I think `strict` is a confusing name. Perhaps `check_value`?

File: query-sql/src/lib.rs [hunk 14]
```diff
@@ -105,9 +105,22 @@ pub enum Constraint {
...
         subquery: TableOrSubquery,
+    },
+    TypeCheck {
+        value: ColumnOrExpression,
+        datatype: SQLDatatype
     }
 }
 
+/// Type safe representation of the possible return values from `typeof`
+pub enum SQLDatatype {
```
Comment (rnewman) [inline query-sql/src/lib.rs]: s/Datatype/TypeAffinity

File: query-translator/src/translate.rs [hunk 15]
```diff
@@ -157,10 +159,53 @@ impl ToConstraint for ColumnConstraint {
                     right: right.into(),
                 }
             },
+            HasTypes { value: table, value_types, strict: strict_requested } => {
+                // If strict mode checking is on, and need to check exactly 1
+                // (not both or neither) of ValueType::Double and ValueType::Long
+                // we emit a Constraint::TypeCheck.
+                let num_numeric_checks = (value_types.contains(ValueType::Double) as i32) +
+                                         (value_types.contains(ValueType::Long) as i32);
+                let strict = strict_requested && num_numeric_checks == 1;
```
Comment (rnewman) [inline query-translator/src/translate.rs]: You're doing the numeric checks even if `!strict_requested`!

So:

```
let strict = strict_requested &&
             value_types.is_unit() &&
             …;
```

Comment (rnewman) [reply query-translator/src/translate.rs]: This bothers me a little, because you're imposing here some knowledge about the precise SQL encoding that we use for types, and that logic is mostly elsewhere. If we add, say, a geospatial floating-point representation — different type tag but `Double` internally — is something going to go wrong here? Can we make this more abstract?

Is the abstract thing you're asking: do we need to check the type affinity of the value to determine whether the value matches our requested value type set?

I think you're saying that for {Double, Integer} we do not need to check the type affinity, because only Double and Integer share a type tag, and if they're both present then the value types correspond to the entire value space.

And so the inverse, which you're doing here, is: are all of the _types_ that are represented by the type tags that are derived from our `value_types` present in `value_types`?

We use `SQLValueTypeSet` to hide all the type-tag-related stuff, so imagine something like:

```
use mentat_core::SQLValueTypeSet;
…

// Ridiculously expensive, but doesn't include its own type tag manipulation.
fn requires_type_affinity_check(value_types: ValueTypeSet) -> bool {
    let mut type_set = ValueTypeSet::none();
    for tag in value_types.iter().map(|vt| vt.value_type_tag()) {
        // from_tag doesn't exist.
        type_set = type_set.union(SQLValueTypeSet::from_tag(tag));
    }
    !type_set.is_strict_subset(value_types)
}
```
    
I'm not suggesting you write exactly that code, but I'd like you to think about what's happening here, and move whatever relates to type tags into `SQLValueType` and friends. This function here should just check a predicate.

File: query-translator/src/translate.rs [hunk 16]
```diff
@@ -157,10 +159,53 @@ impl ToConstraint for ColumnConstraint {
...
+                    // represented using the same value type tag (we choose to
+                    // remove the check for ValueType::Double, but it's an
+                    // arbitrary choice)
+                    value_types.difference(&ValueTypeSet::of_one(ValueType::Double))
+                } else {
+                    value_types
+                };
+                let constraints = types.into_iter().map(|ty| {
+                    let type_column = QualifiedAlias::new(table.clone(), DatomsColumn::ValueTypeTag).to_column();
+                    let loose = Constraint::equal(type_column, ColumnOrExpression::Integer(ty.value_type_tag()));
```
Comment (rnewman) [inline query-translator/src/translate.rs]: `check_type_tag`.

File: query-translator/src/translate.rs [hunk 17]
```diff
@@ -157,10 +159,53 @@ impl ToConstraint for ColumnConstraint {
...
+                        loose
+                    } else {
+                        let val_column = QualifiedAlias::new(table.clone(), DatomsColumn::Value).to_column();
+                        // We're handling strict equality for a numeric type, so we need to emit
+                        // a `typeof(col) = '(real|integer)'` too. This should happen a maximum of
+                        // once per `HasTypes`
+                        Constraint::And {
+                            constraints: vec![
+                                loose,
+                                Constraint::TypeCheck {
```
Comment (rnewman) [inline query-translator/src/translate.rs]: This one would be called `check_type_affinity`.

File: query-translator/src/translate.rs (hunk 15 above)
Comment (thomcc) [reply query-translator/src/translate.rs]: > If we add, say, a geospatial floating-point representation — different type tag but Double internally — is something going to go wrong here?

No, I think that should work. If you added arbitrary precision integer type that had tag = 5 but was stored as a blob, it would be broken though. (Admittedly, you'll loose free comparisons, so I don't think this would be a good idea, but then again, I don't really think using the same tag for Double and Long is a good idea either).

> We use SQLValueTypeSet to hide all the type-tag-related stuff, so imagine something like:

That's still not enough to get to the point where you could add new types that are only discriminated by their SQLite type affinity without modifying translate.rs. Note that a single predicate can never be enough, since you need to know which affinity to emit the check for.

What you end up needing is something like what I've done in the 2nd of my new commits. I'm not sure it's what you had in mind, I'm not a big fan of it since it's more complex (at least conceptually) and less efficient to support cases that can't happen given the way the types are currently set up. I also don't love exporting SQLTypeAffinity from core, but it seems hard to avoid.

File: query-algebrizer/src/clauses/mod.rs (hunk 11 above)
Comment (thomcc) [reply query-algebrizer/src/clauses/mod.rs]: I just went ahead and did this because it bothered me less than the alternative. It also didn't take nearly as long as (for example) the translator changes.

File: query-algebrizer/src/clauses/inputs.rs [hunk 18]
```diff
@@ -30,6 +30,7 @@ use errors::{
 /// the bindings that will be used at execution time.
 /// When built correctly, `types` is guaranteed to contain the types of `values` -- use
 /// `QueryInputs::new` or `QueryInputs::with_values` to construct an instance.
+#[derive(Clone)]
```
[the remark this answers is not in the document]
Comment (thomcc) [reply query-algebrizer/src/clauses/inputs.rs]: This was needed for a test (and that test fails to compile without it). I don't think I agree with your reasoning here, `clone()` is already a known not-cheap action (hence it not being done automatically, like in c++), but it's also your call and I don't care that much so I just changed the test.

**Comment by rnewman** [comment]
98502eb6

Nice work, @thomcc!
